exchanges/populate/src/populateExchange.ts

```diff
@@ -138,21 +141,71 @@ export const extractSelectionsFromQuery = (
   const newFragments: FragmentDefinitionNode[] = [];
   const typeInfo = new TypeInfo(schema);
 
+  const sanitizeSelectionSet = (
+    selectionSet: SelectionSetNode,
+    type: string
+  ) => {
+    const selections: SelectionNode[] = [];
+    const validTypes = (schema.getType(type) as GraphQLObjectType).getFields();
+
+    const validTypeProperties = Object.keys(validTypes);
+
+    selectionSet.selections.forEach(selection => {
+      if (selection.kind === Kind.FIELD) {
+        if (validTypeProperties.includes(selection.name.value)) {
+          if (selection.selectionSet) {
+            selections.push({
+              ...selection,
+              selectionSet: sanitizeSelectionSet(
+                selection.selectionSet,
+                unwrapType(validTypes[selection.name.value].type)!.toString()
+              ),
+            });
+          } else {
+            selections.push(selection);
+          }
+        }
+      } else {
+        selections.push(selection);
+      }
+    });
+    return { ...selectionSet, selections };
+  };
+
   visit(
     query,
     visitWithTypeInfo(typeInfo, {
       Field: node => {
         if (node.selectionSet) {
-          const type = getTypeName(typeInfo);
-          newFragments.push({
-            kind: Kind.FRAGMENT_DEFINITION,
-            typeCondition: {
-              kind: Kind.NAMED_TYPE,
-              name: nameNode(type),
-            },
-            name: nameNode(`${type}_PopulateFragment_`),
-            selectionSet: node.selectionSet,
-          });
+          const type = unwrapType(typeInfo.getType());
+
+          if (isAbstractType(type)) {
+            const types = schema.getPossibleTypes(type);
+            types.forEach(t => {
+              newFragments.push({
+                kind: Kind.FRAGMENT_DEFINITION,
+                typeCondition: {
+                  kind: Kind.NAMED_TYPE,
+                  name: nameNode(t.toString()),
+                },
+                name: nameNode(`${t.toString()}_PopulateFragment_`),
+                selectionSet: sanitizeSelectionSet(
+                  node.selectionSet as SelectionSetNode,
+                  t.toString()
+                ),
+              });
+            });
+          } else if (type) {
+            newFragments.push({
+              kind: Kind.FRAGMENT_DEFINITION,
+              typeCondition: {
+                kind: Kind.NAMED_TYPE,
+                name: nameNode(type.toString()),
```

Comment:
Any reason we need to call `.toString()` on type? Isn't it a string already?

Reply:
It's an instance of the TypeNode class, we did this before as well. Could try and explore different avenues